src/Api/Vault/AuthorizationHandlers/CollectionAccessAuthorizationHandler.cs

```diff
@@ -0,0 +1,112 @@
+    /// Ensures the acting user is allowed to manage access permissions for the target collections.
+    /// </summary>
+    private async Task CanManageCollectionAccessAsync(AuthorizationHandlerContext context, IAuthorizationRequirement requirement, IEnumerable<Guid> collectionIds)
+    {
+        if (!_currentContext.UserId.HasValue)
+        {
+            context.Fail();
+            return;
+        }
+
+        var distinctTargetCollectionIds = collectionIds.Distinct().ToList();
+
+        // List of collections the user is performing the operation on
+        var targetCollections = await _collectionRepository.GetManyByManyIdsAsync(distinctTargetCollectionIds);
+
+        // A target collection does not exist, fail the requirement
+        if (targetCollections.Count != distinctTargetCollectionIds.Count)
+        {
+            context.Fail();
+            return;
+        }
+
+        var userOrgs = await _currentContext.OrganizationMembershipAsync(_organizationUserRepository, _currentContext.UserId.Value);
+        var distinctTargetOrganizationIds = targetCollections.Select(tc => tc.OrganizationId).Distinct();
+        var restrictedOrganizations = new List<CurrentContentOrganization>();
+
+        foreach (var orgId in distinctTargetOrganizationIds)
+        {
+            var org = userOrgs.FirstOrDefault(o => orgId == o.Id);
+
+            // Acting user is not a member of the target organization, fail
+            if (org == null)
+            {
+                context.Fail();
+                return;
+            }
+
+            // Owner/Admins or users with EditAnyCollection permission can always manage collection access
+            if (
+                org.Permissions.EditAnyCollection ||
+                org.Type is OrganizationUserType.Admin or OrganizationUserType.Owner ||
+                await _currentContext.ProviderUserForOrgAsync(org.Id))
+            {
+                continue;
+            }
+
+            restrictedOrganizations.Add(org);
+        }
+
+        // All target collections belong to organizations the acting user is allowed to manage collection access, succeed
+        if (restrictedOrganizations.Count == 0)
+        {
+            context.Succeed(requirement);
+            return;
+        }
+
+        // List of collection Ids the acting user is allowed to manage
+        var manageableCollectionIds =
+            (await _collectionRepository.GetManyByUserIdAsync(_currentContext.UserId.Value))
+            .Where(c => c.Manage)
+            .Select(c => c.Id);
+
+        // Target collections that require explicit "Manage" permission, but the user does not have it
+        var collectionWithoutPermission = from tc in targetCollections
+                                          join o in restrictedOrganizations on tc.OrganizationId equals o.Id
+                                          where !manageableCollectionIds.Contains(tc.Id)
+                                          select tc;
```

Comment:
I don't know if changing to LINQ gets us any performance improvements here, without any change in the underlying logic. Is join more efficient than a loop?

I went down a rabbit hole here, I thought I should work it through myself if I was going to ask for more changes, and here's what I came up with: https://github.com/bitwarden/server/compare/ac/ac-1174/collection-user-auth-handler...ac/ac-1174/collection-user-auth-handler-alt

However I'm not convinced that it's necessarily better, it might just be different. What do you think? Your choice.

If you want to stick with the LINQ approach then my only recommendation is to use a hashset for constant time lookup on `manageableCollectionIds`.

```cs
        var manageableCollectionIds =
            (await _collectionRepository.GetManyByUserIdAsync(_currentContext.UserId.Value))
            .Where(c => c.Manage)
            .Select(c => c.Id)
            .ToHashSet();
```

Reply:
Thanks for digging into this; however, I went ahead and simplified this logic by requiring that all of the resources belong to the same organization. I did still add the `ToHashSet()` call to improve the lookup time on `managableCollectionIds`, but the other nested loops are no longer a problem this way.

Supporting multiple orgs in a single request has been something I kept going back and forth on. It adds additional complexity and is for something we currently have no use cases for. Users never modify resources belonging to multiple organizations at once.

Additionally, as I have been working on the the command/SQL to actually create these associations, trying to support multiple organizations blows up the complexity of the SQL and validation logic; so it seemed best to just focus on our actual use case and restrict requests to have the same organization for all collections.

Let me know what you think!

Reply:
I think your reasoning is sound, especially given that we don't actually need that extra complexity. (and we're unlikely to need it in the future)